internal/worker/jobworker/jobworker.go

```diff
@@ -0,0 +1,186 @@
+// The provided context will kill the process if it becomes done, so don't use
+// an ephemeral context like one for a particular request. The userID is an
+// opaque value that is used for authorization of later requests. Only matching
+// userIDs will be able to Stop or get the Status or Output of a job. Returns
+// the opaque job.ID that is required for subsequent operations with the job.
+func (w *JobWorker) StartJob(ctx context.Context, userID user.ID, command string, args ...string) (job.ID, error) {
+	cmdArgs := w.reexec.Args
+	cmdArgs = append(cmdArgs, command)
+	cmdArgs = append(cmdArgs, args...)
+
+	j, err := job.New(
+		ctx,
+		w.cfg,
+		userID,
+		w.reexec.Command,
+		cmdArgs,
+		w.reexec.Env,
+	)
+	if err != nil {
+		return job.ID{}, err
+	}
+
+	if err = j.Start(); err != nil {
+		return job.ID{}, err
+	}
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	w.jobs[j.ID()] = j
+
+	return j.ID(), nil
+}
+
+// StartJobChild is called when this binary is reexecuted with the new
+// namespaces applied. It should be the only method called by the binary in that
+// circumstance and should never be called in any other situation. It will
+// create a new cgroup with cpu, memory and io limits applied, then it will
+// remount /proc and finally it will execute the command, with optional args.
+func (w *JobWorker) StartJobChild(_ context.Context, command string, args ...string) error {
+	cmd, err := exec.LookPath(command)
+	if err != nil {
+		err = fmt.Errorf("LookPath error: %w", err)
+		fmt.Fprintln(os.Stderr, err)
+		return err
+	}
+
+	if err = w.prepJobEnv(); err != nil {
+		err = fmt.Errorf("prepJobEnv error: %w", err)
+		fmt.Fprintln(os.Stderr, err)
+		return err
+	}
+
+	args = append([]string{cmd}, args...)
+	if err = syscall.Exec(cmd, args, os.Environ()); err != nil {
+		err = fmt.Errorf("syscall.Exec error: %w", err)
+		fmt.Fprintln(os.Stderr, err)
+		return err
+	}
+
+	return nil
+}
+
+// ErrJobNotFound is returned when trying to stop, get status or get output of a
+// job that doesn't exist or that the user is not authorized for.
+var ErrJobNotFound = errors.New("job not found")
+
+func (w *JobWorker) getJob(userID user.ID, jobID job.ID) (*job.Job, error) {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	job, ok := w.jobs[jobID]
+	if !ok || job.UserID() != userID {
+		return nil, ErrJobNotFound
+	}
+
+	return job, nil
+}
+
+// StopJob kills the job identified by jobID. If the job does not exist, or if
+// the user is not authorized, ErrJobNotFound will be returned.
+func (w *JobWorker) StopJob(_ context.Context, userID user.ID, jobID job.ID) error {
+	job, err := w.getJob(userID, jobID)
+	if err != nil {
+		return err
+	}
+
+	return job.Stop()
+}
+
+// JobStatus will return the JobStatus and optional ExitCode from the job. The
+// ExitCode will not exist if the job is still running, or if it exited
+// uncleanly, e.g. because it was killed by a signal. If the job does not exist,
```

Comment:
> // ExitCode will not exist (...) if it exited uncleanly, e.g. because it was killed by a signal.

Is this correct?

Reply:
I thought it was when I wrote it, then my testing showed that it wasn't. I'll update.